File: Problemset/add-two-numbers/add-two-numbers..py

# @Title: 两数相加 (Add Two Numbers)
# @Author: KivenC
# @Date: 2018-12-08 14:04:48
# @Runtime: 140 ms
# @Memory: 7.1 MB

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        result = ListNode(0)
        cur = result
        while l1 or l2:
            cur.val += self.addNode(l1, l2)
            if cur.val > 9:
                cur.val -= 10
                cur.next = ListNode(1)
            else:
                if l1 and l1.next or l2 and l2.next:
                    cur.next = ListNode(0)
            cur = cur.next
            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next
        return result
    
    def addNode(self, node1, node2):
        if not node1 and not node2:
            pass
        if not node1:
            return node2.val
        if not node2:
            return node1.val
        return node1.val + node2.val

File: Problemset/add-two-numbers/add-two-numbers.go

// @Title: 两数相加 (Add Two Numbers)
// @Author: KivenC
// @Date: 2019-06-05 10:41:01
// @Runtime: 16 ms
// @Memory: 5.2 MB

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
// 	// way 1
// 	var res []ListNode
// 	node1, node2 := l1, l2
// 	tmp := 0
// 	for {
// 		if node1 != nil {
// 			tmp += node1.Val
// 			node1 = node1.Next
// 		}
// 		if node2 != nil {
// 			tmp += node2.Val
// 			node2 = node2.Next
// 		}
// 		res = append(res, ListNode{Val: tmp % 10})
// 		tmp = tmp / 10
// 		if node1 == nil && node2 == nil && tmp == 0 {
// 			break
// 		}
// 	}
// 	for i := len(res) - 1; i > 0; i-- {
// 		res[i-1].Next = &res[i]
// 	}
// 	return &res[0]
    
    // way 2
    resPre := &ListNode{}
	cur := resPre
	carry := 0

	for l1 != nil || l2 != nil || carry > 0 {
		sum := carry

		if l1 != nil {
			sum += l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			sum += l2.Val
			l2 = l2.Next
		}

		carry = sum / 10

		cur.Next = &ListNode{Val: sum % 10}
		cur = cur.Next
	}
	return resPre.Next   
}

